tools/perf/ResultsComparer/Bdn/BdnComparer.cs

```diff
@@ -0,0 +1,128 @@
+                return Task.FromResult(results);
+            }
+
+            results.Results = notSame.Select(r =>
+                new ComparerResult { Id = r.id, BaseResult = r.baseResult, DiffResult = r.diffResult, Conclusion = r.conclusion });
+
+            return Task.FromResult(results);
+        }
+
+        private static IEnumerable<(string id, Benchmark baseResult, Benchmark diffResult, EquivalenceTestConclusion conclusion)> GetNotSameResults(string basePath, string diffPath, ComparerOptions options, Threshold testThreshold, Threshold noiseThreshold)
+        {
+            foreach ((string id, Benchmark baseResult, Benchmark diffResult) in ReadResults(basePath, diffPath, options)
+                .Where(result => result.baseResult.Statistics != null && result.diffResult.Statistics != null)) // failures
+            {
+                var baseValues = baseResult.GetOriginalValues();
+                var diffValues = diffResult.GetOriginalValues();
+
+                var userTresholdResult = StatisticalTestHelper.CalculateTost(MannWhitneyTest.Instance, baseValues, diffValues, testThreshold);
+                if (userTresholdResult.Conclusion == EquivalenceTestConclusion.Same)
+                    continue;
+
+                var noiseResult = StatisticalTestHelper.CalculateTost(MannWhitneyTest.Instance, baseValues, diffValues, noiseThreshold);
+                if (noiseResult.Conclusion == EquivalenceTestConclusion.Same)
+                    continue;
+
+                yield return (id, baseResult, diffResult, userTresholdResult.Conclusion);
+            }
+        }
+
+        private static IEnumerable<(string id, Benchmark baseResult, Benchmark diffResult)> ReadResults(string basePath, string diffPath, ComparerOptions options)
+        {
+            var baseFiles = GetFilesToParse(basePath);
+            var diffFiles = GetFilesToParse(diffPath);
+
+            if (!baseFiles.Any() || !diffFiles.Any())
+                throw new ArgumentException($"Provided paths contained no {FullBdnJsonFileExtension} files.");
+
+            var baseResults = baseFiles.Select(ReadFromFile);
+            var diffResults = diffFiles.Select(ReadFromFile);
+
+            var filters = options.Filters.Select(pattern => new Regex(WildcardToRegex(pattern), RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)).ToArray();
+
+            var benchmarkIdToDiffResults = diffResults
+                .SelectMany(result => result.Benchmarks)
+                .Where(benchmarkResult => !filters.Any() || filters.Any(filter => filter.IsMatch(benchmarkResult.FullName)))
+                .ToDictionary(benchmarkResult => benchmarkResult.FullName, benchmarkResult => benchmarkResult);
+
+            return baseResults
+                .SelectMany(result => result.Benchmarks)
+                .ToDictionary(benchmarkResult => benchmarkResult.FullName, benchmarkResult => benchmarkResult) // we use ToDictionary to make sure the results have unique IDs

```

Comment:
nit: Wouldn't Distinct be better here?

Reply:
In this case we expect all the results to have unique names. If there are duplicates, an exception will be thrown. It's not a grouping of distinct values.